### ISSUE TITLE
🐞 Bug Report: Accordion and Toggle have broken accessibility

### ISSUE THREAD

**Issue by clmnsm** [elementor/elementor#23335]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

Both toggle and accordion (non-container variant) use aria-attributes like -expanded and -controls, which is great. 

Unfortunately, they also misuse tabindex which make those attributes invisible to users that rely on keyboard navigation.

### Steps to reproduce

❌ Opened items have both the "elementor-tab-title" and the "elementor-accordion-title/elementor-toggle-title" at tabindex 0. 
✔️ Only "elementor-tab-title" should be at tabindex 0, "elementor-accordion-title/elementor-toggle-title" should be at -1 because it doesn't provide meaningful aria info – it's just an empty link. 

❌ Closed items habe only "elementor-accordion-title/elementor-toggle-title" at tabindex 0. The "elementor-tab-title" is at -1 and therefore not focusable.
✔️ Instead, the "elementor-tab-title" should be at 0, because it provides meaningful information like aria-expanded="false".

### Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [X] I can reproduce this bug consistently following the steps above.

### System Info

irrelevant

**Comment by nicholaszein** [comment]
Hi again, @clmnsm! 

Thank you for submitting your issue! 🙏 

Please allow us some time to investigate it, and get back to you with our findings. 🧐 

Best regards

**Comment by hein-obox** [comment]
@nicholaszein Accessibility still needs to be implemented for the new widget.

**Comment by clmnsm** [comment]
The new nested accordion seems to be fine, actually. It's using html native summary/details which at least ChromeVox interprets correctly. 

If only there was a conversion button... 
Manual copying of accordions with dozens of entries is a real chore.

**Comment by renzat** [comment]
On my site the Elementor toggle module wouldn't open on windows without a WordPress admin bar at the top. Switching from Hello theme to GeneratePress theme fixed the issue.

Hello Theme / Elementor / Elementor Pro no WordPress admin bar (toggle does not open):

<div id="elementor-tab-title-1401" class="elementor-tab-title" data-tab="1" role="button" aria-controls="elementor-tab-content-1401" aria-expanded="false">

Hello Theme / Elementor / Elementor Pro WITH WordPress admin bar (toggle opens):

<div id="elementor-tab-title-1401" class="elementor-tab-title" data-tab="1" role="button" aria-controls="elementor-tab-content-1401" aria-expanded="false" tabindex="-1" aria-selected="false">

GeneratePress / Elementor Elementor Pro WITHOUT WordPress admin bar (toggle opens):

<div id="elementor-tab-title-1401" class="elementor-tab-title" data-tab="1" role="button" aria-controls="elementor-tab-content-1401" aria-expanded="false" tabindex="-1" aria-selected="false">

**Comment by mateuswetah** [comment]
Not sure if related, but the `<a class="elementor-accordion-title"` definitely needs a `role="button"`. This is also causing SEO issues on the Google Lighthouse Report.

**Comment by manishagarwal-tech** [comment]
Also using elementor accordion it breaks the accessibility, as it generates aria-selected and aria-expnaded to true on active accordion.

Generated DOM:

<div id="elementor-tab-title-7471" class="elementor-tab-title elementor-active" data-tab="1" role="button" aria-controls="elementor-tab-content-7471" aria-expanded="true" tabindex="0" aria-selected="true">


Issue: ARIA attribute is not allowed: aria-selected="true"

How can we remove or disallow aria-selected should not rendered.

Any suggestion on it?

**Comment by hein-obox** [comment]
@nicholaszein I believe we should redo the accessibility implementation of the Nested Accordion and the Nested Tabs, to remove the keyboard navigation and to replace it with the tab navigation, as we have for the Mega Menu. I have opened a task for that.

In the implementation we should add Lighthouse tests inside our Playwright tests, to make sure we comply with all Lighthouse guidelines.

@manishagarwal-tech I am not sure what you mean by 'ARIA attribute is not allowed: aria-selected="true"'. What do you mean by that? Is there a Lighthouse violation? Or what makes you say that the aria attribute isn't allowed here?

**Comment by manishagarwal-tech** [comment]
Hey @hein-obox 
We have used the elementor accordion on our website to display some FAQs, when we run the accessibility test on that page, we are getting warning about the accordion DOM (generated by elementor itself), We are using axeDevTools chrome addon to check the accessibility improvements.

I am not sure this issue is not being recreated on the lighthouse. can you guide me why I am getting that issue while using elementor.

DOM Generated by Elementor:

<div id="elementor-tab-title-7471" class="elementor-tab-title elementor-active" data-tab="1" role="button" aria-controls="elementor-tab-content-7471" aria-expanded="true" tabindex="0" aria-selected="true">


Warning (by: axeDevTools):

ARIA attribute is not allowed: aria-selected="true"

It says about more information you may check https://dequeuniversity.com/rules/axe/4.8/aria-allowed-attr?application=AxeChrome

It is something need to be fix at elementor end or my end, please suggest.

**Comment by manishagarwal-tech** [comment]
> DOM Generated by Elementor:

`<div id="elementor-tab-title-7471" class="elementor-tab-title elementor-active" data-tab="1" role="button" aria-controls="elementor-tab-content-7471" aria-expanded="true" tabindex="0" aria-selected="true">
`

**Comment by manishagarwal-tech** [comment]
> > DOM Generated by Elementor:
> 
> `<div id="elementor-tab-title-7471" class="elementor-tab-title elementor-active" data-tab="1" role="button" aria-controls="elementor-tab-content-7471" aria-expanded="true" tabindex="0" aria-selected="true"> `

and for non-active accordion both `aria-selected` and `aria-expanded` set to `false`

**Comment by hein-obox** [comment]
Thank you, @manishagarwal-tech for your feedback. 

@doronwolf We need to prioritise the Nested Accordion task, because there seems to a problem currently.

**Comment by ranjeetananda** [comment]
@hein-obox for the issue reported by @manishagarwal-tech - do you have an eta on the fix on the elementor side? This is causing accessibility test to fail on the pages where accordion is used. 

![image](https://github.com/elementor/elementor/assets/50963727/b9237d1e-1a0c-401a-96ba-1312327c6fec)

**Comment by hein-obox** [comment]
@ranjeetananda @manishagarwal-tech From what I can see this is an issue inside the old Accordion widget. The quickest solution might be to start using the new nested widget, available under the Nested Elements experiment. See https://elementor.com/help/accordion-widget/#:~:text=If%20you%20are%20using%20the,and%20buttons%20to%20your%20Accordions (bottom of the page).

@nicholaszein @rami-elementor Can this be solved in the existing Accordion widget as well?

**Comment by ranjeetananda** [comment]
Is there a way to fix this in existing Accordion Widget. Recreating the pages with new nested accordions is going to be a lot of work for us.

**Comment by KennethErvinYoung** [comment]
**Toggle Element Page Speed Insights SEO Bug**

I verified that the DaveDen2 code resolves the problem on the Toggle Element.
I tried the PHP code from sebsonic2o, but it did NOT work on the Toggle element. It's been ten months. Perhaps Elementor changed something that broke his code.
I will modify the DaveDen2 code to try to fix the Accordion element, which is failing Google PSI for "aria—* attributes do not match their roles." This may be the same issue or a similar one. Please share if anyone has successfully implemented a PHP code snippet for the Accordion element.

<?php

add_filter( 'elementor/frontend/the_content', 'add_role_attribute_to_elementor_links', 10, 2 );

function add_role_attribute_to_elementor_links( $content ) {
    $dom = new DOMDocument();
    $dom->loadHTML( mb_convert_encoding( $content, 'HTML-ENTITIES', 'UTF-8' ) );
    $xpath = new DOMXPath( $dom );
    $elements = $xpath->query( "//a[contains(concat(' ', normalize-space(@class), ' '), ' elementor-accordion-title ') or contains(concat(' ', normalize-space(@class), ' '), ' elementor-toggle-title ')]" );
    if ( $elements->length > 0 ) {
        foreach ( $elements as $element ) {
            $element->setAttribute( 'role', 'presentation' );
        }
        $content = $dom->saveHTML();
    }
    return $content;
}

**Accordion Bug - Page Speed Insights Accessibility Bug:**

These guys claim they have verified the code identifying the Elementor Accordion coding error, and it only needs a simple one-line fix. I haven't tried it yet, but I'm going to.

https://www.testpros.com/accessibility/making-elementor-essential-addons-accordion-accessible/

**Comment by Trott** [comment]
I'm pretty sure the problem is this code which sets `aria-selected` regardless of whether the tab title has `role=button` or not. (This is in `activateTab()`. There is a similar issue a few lines above that in the `deactivateTab()`.)

https://github.com/elementor/elementor/blob/c4a9dc5fd2d452c7e7eac1eecb9683e9bd397e52/assets/dev/js/frontend/handlers/base-tabs.js#L132-L136

**Comment by patlewpl** [comment]
did you fix this issue somehow?

**Comment by Trott** [comment]
> did you fix this issue somehow?

It's been a while since I've looked at this issue, but I *think* these patches that I float on our Elementor installation are intended to fix this. It might not be the "right" fix and it might not resolve the issue for all use cases, but it resolves the issue for me.

Fix for toggle:

```patch
diff --git a/wp-content/plugins/elementor/includes/widgets/toggle.php b/wp-content/plugins/elementor/includes/widgets/toggle.php
index 2035b487a..05d493dfe 100644
--- a/wp-content/plugins/elementor/includes/widgets/toggle.php
+++ b/wp-content/plugins/elementor/includes/widgets/toggle.php
@@ -643,7 +643,7 @@ class Widget_Toggle extends Widget_Base {
 							<?php } ?>
 						</span>
 						<?php endif; ?>
-						<a class="elementor-toggle-title"><?php $this->print_unescaped_setting( 'tab_title', 'tabs', $index ); ?></a>
+						<a class="elementor-toggle-title" tabindex="0"><?php $this->print_unescaped_setting( 'tab_title', 'tabs', $index ); ?></a>
 					</<?php Utils::print_validated_html_tag( $settings['title_html_tag'] ); ?>>
 
 					<div <?php $this->print_render_attribute_string( $tab_content_setting_key ); ?>><?php Utils::print_unescaped_internal_string( $this->parse_text_editor( $item['tab_content'] ) ); ?></div>
@@ -730,7 +730,7 @@ class Widget_Toggle extends Widget_Base {
 								<# } #>
 							</span>
 							<# } #>
-							<a class="elementor-toggle-title">{{{ item.tab_title }}}</a>
+							<a class="elementor-toggle-title" tabindex="0">{{{ item.tab_title }}}</a>
 						</{{{ titleHTMLTag }}}>
 						<div {{{ view.getRenderAttributeString( tabContentKey ) }}}>{{{ item.tab_content }}}</div>
 					</div>
```

Fix for accordion:

```patch
diff --git a/wp-content/plugins/elementor/includes/widgets/accordion.php b/wp-content/plugins/elementor/includes/widgets/accordion.php
index e6eb3a331..fc290d8ae 100644
--- a/wp-content/plugins/elementor/includes/widgets/accordion.php
+++ b/wp-content/plugins/elementor/includes/widgets/accordion.php
@@ -617,7 +617,7 @@ class Widget_Accordion extends Widget_Base {
 							<?php } ?>
 							</span>
 						<?php endif; ?>
-						<a class="elementor-accordion-title"><?php
+						<a class="elementor-accordion-title" tabindex="0"><?php
 							$this->print_unescaped_setting( 'tab_title', 'tabs', $index );
 						?></a>
 					</<?php Utils::print_validated_html_tag( $settings['title_html_tag'] ); ?>>
@@ -709,7 +709,7 @@ class Widget_Accordion extends Widget_Base {
 								<# } #>
 							</span>
 							<# } #>
-							<a class="elementor-accordion-title">{{{ item.tab_title }}}</a>
+							<a class="elementor-accordion-title" tabindex="0">{{{ item.tab_title }}}</a>
 						</{{{ titleHTMLTag }}}>
 						<div {{{ view.getRenderAttributeString( tabContentKey ) }}}>{{{ item.tab_content }}}</div>
 					</div>
```
